### ISSUE TITLE
Cisco ASA Decoders

### ISSUE THREAD

**Issue by soulmaster43** [wazuh/wazuh-ruleset#556]
The Cisco ASA decoder for denied icmp messages requires some improvement as the current one doesn't parse the events properly (wrong order for fields)
![Capture](https://user-images.githubusercontent.com/19363009/73637778-56cce380-4669-11ea-8f1d-3cd97982d281.PNG)

Also, I am suggesting renaming the fields src_ip, src_port, dst_ip, dst_port to srcip, srcport, dstip, dstport as these fields are consistent with all the other decoders.

These are of course minor improvements but I feel they are worth the effort.

**Comment by Zenidd** [comment]
Hello @soulmaster43,

First of all, thanks for your contribution. I hope you don't mind I moved the issue to the **wazuh-ruleset** repository.

> The Cisco ASA decoder for denied icmp messages requires some improvement as the current one doesn't parse the events properly (wrong order for fields)

- The fields for that log are indeed incorrectly decoded. I first thought the event above was yours but it is [our sample log for the 4-313004 ID message in the Cisco ASA decoders file](https://github.com/wazuh/wazuh/blob/61af23590aa2df6e069f35c12d06641091805cdf/etc/decoders/0064-cisco-asa_decoders.xml#L461-L469).


```XML
<!--
%ASA-4-313004: Denied ICMP type=0, from laddr 80.241.208.43 on interface outside to 116.6.127.116: no matching session
-->
<decoder name="cisco-asa-denied-icmp">
    <parent>cisco-asa</parent>
    <prematch offset="after_parent">4-313004</prematch>
    <regex offset="after_parent">(\w+): (\.+) type=(\S+), from laddr (\S+) on interface (\S+) to (\S+): (\.+)</regex>
    <order>id, description, src_ip, interface, dst_ip, reason</order>
</decoder>
```

The decoder extracts 7 fields, one for each use of the `(` `)` operator. However, there are just 6 target variables `<order>id, description, src_ip, interface, dst_ip, reason</order>`. The variable that should receive the `type` value is missing so the other values have shifted to the variable from before. Adding the `type` variable in the `<order>` tag is enough to make the decoder work properly.

```XML
<decoder name="cisco-asa-denied-icmp">
    <parent>cisco-asa</parent>
    <prematch offset="after_parent">4-313004</prematch>
    <regex offset="after_parent">(\w+): (\.+) type=(\S+), from laddr (\S+) on interface (\S+) to (\S+): (\.+)</regex>
    <order>id, description, type, src_ip, interface, dst_ip, reason</order>
</decoder>
```



```
**Phase 1: Completed pre-decoding.
       full event: '%ASA-4-313004: Denied ICMP type=0, from laddr 80.241.208.43 on interface outside to 116.6.127.116: no matching session'
       timestamp: '(null)'
       hostname: 'nanoPill'
       program_name: '(null)'
       log: '%ASA-4-313004: Denied ICMP type=0, from laddr 80.241.208.43 on interface outside to 116.6.127.116: no matching session'

**Phase 2: Completed decoding.
       decoder: 'cisco-asa'
       id: '4-313004'
       description: 'Denied ICMP'
       type: '0'
       src_ip: '80.241.208.43'
       interface: 'outside'
       dst_ip: '116.6.127.116'
       reason: 'no matching session'

**Phase 3: Completed filtering (rules).
       Rule id: '64004'
       Level: '3'
       Description: 'ASA warning message.'
**Alert to be generated.
```

> Also, I am suggesting renaming the fields src_ip, src_port, dst_ip, dst_port to srcip, srcport, dstip, dstport as these fields are consistent with all the other decoders.

- I think these variables were named with these bit different names to turn them into [dynamic fields](https://documentation.wazuh.com/3.11/user-manual/ruleset/dynamic-fields.html#dynamic-decoders). This way, [same_field](https://documentation.wazuh.com/3.11/user-manual/ruleset/ruleset-xml-syntax/rules.html#same-field) and [not_same_fields](https://documentation.wazuh.com/3.11/user-manual/ruleset/ruleset-xml-syntax/rules.html#not-same-field) rules options can be used over the decoded fields. Maybe we should rethink whether it's more useful for these fields to be static or dynamic.

I'm creating the PR with the `<order>` option fix. Thanks again for your contribution.

Greetings, JP Sáez

**Comment by soulmaster43** [comment]
Hi JP,

Thank you for picking this up.
Since we're on the topic of Cisco ASA decoders, I would like to share the decoders file I'm using.
Has 2-3 more decoders aside from the wazuh one. Perhaps you may find it useful.

Best,

Alex

**Comment by vikman90** [comment]
@soulmaster43 @Zenidd this issue has been fixed by PR #558.

Thank you all!
